src/nvim/os/pty_process_unix.c

```diff
@@ -273,26 +273,24 @@ static void chld_handler(uv_signal_t *handle, int signum)
   int stat = 0;
   int pid;
 
-  do {
-    pid = waitpid(-1, &stat, WNOHANG);
-  } while (pid < 0 && errno == EINTR);
-
-  if (pid <= 0) {
-    return;
-  }
-
   Loop *loop = handle->loop->data;
 
   kl_iter(WatcherPtr, loop->children, current) {
     Process *proc = (*current)->data;
-    if (proc->pid == pid) {
-      if (WIFEXITED(stat)) {
-        proc->status = WEXITSTATUS(stat);
-      } else if (WIFSIGNALED(stat)) {
-        proc->status = WTERMSIG(stat);
-      }
-      proc->internal_exit_cb(proc);
-      break;
+    do {
+      pid = waitpid(proc->pid, &stat, WNOHANG);
+    } while (pid < 0 && errno == EINTR);
+
+    if (pid <= 0) {
+      continue;
+    }
+
+    if (WIFEXITED(stat)) {
+      proc->status = WEXITSTATUS(stat);
+    } else if (WIFSIGNALED(stat)) {
+      proc->status = WTERMSIG(stat);
     }
+    proc->internal_exit_cb(proc);
+    break;
```

Comment:
What if three or more children are terminated? Would the first signal be handled here, the second signal be marked pending and every further (of the same) signal be dropped (we miss a child termination)? Would that be fixed if we remove the `break` here?

Reply:
Although in both the original code and this PR we only process one job, I can understand oni-link's concern that we may be handling the wrong job because we end up reaping a different child than the one that delivered the signal.

In order to truly be deterministic, though, we would likely need to use `sigaction()` so we can be directly told which child delivered the signal.  Since we're using libuv's abstraction, that's not really an option here.

I wonder if a simpler fix is to change the original code to use `waitpid(0, &stat, WNOHANG);`.  It looks like that should still fix the problem.  If that does work, we probably want to actually loop until `waitpid()` tells us none of our children are waitable.

Reply:
If I understand correctly the concern is not that we handle jobs in the 'wrong' order - that really shouldn't matter.  (A dies, B dies, we handle B, we handle A - so what?  B might just as well have died first).  

Rather the worry was that we might somehow *miss* terminations.  I don't think that'll happen, though I'm not 100% sure and I can't see either how removing the `break` would be unsafe.  (But I still say, if you want to do that, let's do it in a new issue).

Either way, I'd favour sticking with being completely explicit about what PIDs we're waiting for as the most foolproof approach.

Reply:
I've done some more googling and I now think that @oni-link is right all along - `SIGCHLD`s are not queued, so we might see fewer of them than terminated processes.  

I'll raise a new issue and propose that the fix is simply to remove the `break`